src/main/java/com/aws/greengrass/logmanager/services/DiskSpaceManagementService.java

```diff
@@ -5,214 +5,48 @@
-                                updatedComponentsConfiguration.values()) {
-                            try {
-                                deleteFilesIfNecessary(componentLogConfiguration);
-                            } catch (UncheckedIOException e) {
-                                logger.atWarn().log("Unchecked error thrown when collecting files to be deleted",
-                                        Utils.getUltimateCause(e));
-                            }
-                        }
-                    }
-
-                    if (!watchKey.reset()) {
-                        logger.atError().log("Log Space Management encountered an issue. Returning.");
-                        scheduleSpaceManagementThread(componentLogConfigurations);
-                        break;
-                    }
-                }
-            } catch (IOException e) {
-                // If there is any other IOException, then we should restart the thread.
-                scheduleSpaceManagementThread(componentLogConfigurations);
-            }
+    public void freeDiskSpace(LogFileGroup group, Instant lastUpdated) {
+        if (!group.hasExceededDiskUsage()) {
+            return;
         }
-    }
 
-    /**
-     * Deletes the oldest log files which matches the file name pattern. The log files will be deleted until the
-     * minimum number of bytes have been deleted.
-     *
-     * @param componentLogConfiguration Log configuration to investigate for deletion
-     * @throws IOException for errors during directory walking
-     */
-    private void deleteFilesIfNecessary(ComponentLogConfiguration componentLogConfiguration) throws IOException {
-        if (componentLogConfiguration.getDiskSpaceLimit() == null) {
+        if (!group.getMaxBytes().isPresent()) {
             return;
         }
-        try (Stream<Path> fileStream = Files.walk(componentLogConfiguration.getDirectoryPath())
-                .filter(p -> {
-                    File file = p.toFile();
-                    return file.isFile() && componentLogConfiguration.getFileNameRegex()
-                            .matcher(file.getName()).find();
-                })) {
-            List<Path> paths = fileStream.collect(Collectors.toList());
-            long totalBytes = paths.stream().mapToLong(p -> p.toFile().length()).sum();
-            long minimumBytesToBeDeleted = totalBytes - componentLogConfiguration.getDiskSpaceLimit();
 
-            // If we don't need to remove any bytes, or if the file count is only 1 (or less), then there's nothing
-            // to do.
-            if (minimumBytesToBeDeleted <= 0 || paths.size() < 2) {
-                return;
+        long bytesDeleted = 0;
+        long minimumBytesToBeDeleted = Math.max(group.totalSizeInBytes() - group.getMaxBytes().get(), 0);
+
+        List<LogFile> deletableFiles = group.getLogFiles().stream()
+                .filter(file ->  lastUpdated.isAfter(Instant.ofEpochMilli(file.lastModified())))
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;lastUpdated&#39; (index: 1 | type: Instant). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;group&#39;. The same parameter type is validated here for example: aws-greengrass-log-manager/src/main/java/com/aws/greengrass/logmanager/model/LogFileGroup.java:106. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).